crates/miden-batch-prover/src/local_batch_prover.rs

```diff
@@ -0,0 +1,682 @@
+
+        let note0 = mock_output_note(50);
+        let tx1 =
+            MockProvenTxBuilder::with_account(account1.id(), Digest::default(), account1.hash())
+                .output_notes(vec![note0.clone()])
+                .build()?;
+        let tx2 =
+            MockProvenTxBuilder::with_account(account2.id(), Digest::default(), account2.hash())
+                .output_notes(vec![note0.clone()])
+                .build()?;
+
+        let error = LocalBatchProver::prove(ProposedBatch::new(
+            [tx1.clone(), tx2.clone()].into_iter().map(Arc::new).collect(),
+            NoteInclusionProofs::default(),
+        ))
+        .unwrap_err();
+
+        assert_matches!(error, BatchError::DuplicateOutputNote {
+          note_id,
+          first_transaction_id,
+          second_transaction_id
+        } if note_id == note0.id() &&
+          first_transaction_id == tx1.id() &&
+          second_transaction_id == tx2.id());
+
+        Ok(())
+    }
+
+    /// Tests that a note created and consumed in the same batch are erased from the input and
+    /// output note commitments.
+    #[test]
+    fn note_created_and_consumed_in_same_batch() -> anyhow::Result<()> {
+        let account1 = mock_wallet_account(10);
+        let account2 = mock_wallet_account(100);
+
+        let note0 = mock_note(50);
+        let tx1 =
+            MockProvenTxBuilder::with_account(account1.id(), Digest::default(), account1.hash())
+                .output_notes(vec![OutputNote::Full(note0.clone())])
+                .build()?;
+        let tx2 =
+            MockProvenTxBuilder::with_account(account2.id(), Digest::default(), account2.hash())
+                .unauthenticated_notes(vec![note0.clone()])
+                .build()?;
+
+        let batch = LocalBatchProver::prove(ProposedBatch::new(
+            [tx1, tx2].into_iter().map(Arc::new).collect(),
+            NoteInclusionProofs::default(),
+        ))?;
+
+        assert_eq!(batch.input_notes().len(), 0);
+        assert_eq!(batch.output_notes().len(), 0);
+        assert_eq!(batch.output_notes_tree().num_leaves(), 0);
+
+        Ok(())
+    }
+
+    /// Test that an authenticated input note that is also created in the same batch does not error
+    /// and instead is marked as consumed.
+    /// - This requires a nullifier collision on the input and output note which is very unlikely in
+    ///   practice.
+    /// - This makes the created note unspendable as its nullifier is added to the nullifier tree.
+    /// - The batch kernel cannot return an error in this case as it can't detect this condition due
+    ///   to only having the nullifier for authenticated input notes _but_ not having the nullifier
+    ///   for private output notes.
+    /// - We test this to ensure the kernel does something reasonable in this case and it is not an
+    ///   attack vector.
+    #[test]
+    fn authenticated_note_created_in_same_batch() -> anyhow::Result<()> {
```

Comment:
This is a pretty tricky case I want to call out. This describes the best way I can think of to handle it and the rationale, but not sure if there's a better way.